src/Resources/contao/library/Contao/User.php

```diff
@@ -258,481 +279,475 @@ public function getData()
-			if (isset($GLOBALS['TL_HOOKS']['importUser']) && \is_array($GLOBALS['TL_HOOKS']['importUser']))
-			{
-				foreach ($GLOBALS['TL_HOOKS']['importUser'] as $callback)
-				{
-					$this->import($callback[0], 'objImport', true);
-					$blnLoaded = $this->objImport->{$callback[1]}(\Input::post('username', true), $request->request->get('password'), $this->strTable);
-
-					// Load successfull
-					if ($blnLoaded === true)
-					{
-						break;
-					}
-				}
-			}
+		return false;
+	}
 
-			// Return if the user still cannot be loaded
-			if (!$blnLoaded || $this->findBy('username', \Input::post('username', true)) == false)
-			{
-				\Message::addError($GLOBALS['TL_LANG']['ERR']['invalidLogin']);
-				$this->log('Could not find user "' . \Input::post('username', true) . '"', __METHOD__, TL_ACCESS);
 
-				return false;
-			}
+	/**
+	 * Return true if the user is member of a particular group
+	 *
+	 * @param integer $id The group ID
+	 *
+	 * @return boolean True if the user is a member of the group
+	 */
+	public function isMemberOf($id)
+	{
+		// ID not numeric
+		if (!is_numeric($id))
+		{
+			return false;
 		}
 
-		$time = time();
+		$groups = \StringUtil::deserialize($this->arrData['groups']);
 
-		// Set the user language
-		if (\Input::post('language'))
+		// No groups assigned
+		if (empty($groups) || !is_array($groups))
 		{
-			$this->language = \Input::post('language');
+			return false;
 		}
 
-		// Lock the account if there are too many login attempts
-		if ($this->loginCount < 1)
+		// Group ID found
+		if (in_array($id, $groups))
 		{
-			$this->locked = $time;
-			$this->loginCount = \Config::get('loginCount');
-			$this->save();
+			return true;
+		}
 
-			// Add a log entry and the error message, because checkAccountStatus() will not be called (see #4444)
-			$this->log('User "' . $this->username . '" has been locked for ' . ceil(\Config::get('lockPeriod') / 60) . ' minutes', __METHOD__, TL_ACCESS);
-			\Message::addError(sprintf($GLOBALS['TL_LANG']['ERR']['accountLocked'], ceil((($this->locked + \Config::get('lockPeriod')) - $time) / 60)));
+		return false;
+	}
 
-			// Send admin notification
-			if (\Config::get('adminEmail') != '')
-			{
-				$objEmail = new \Email();
-				$objEmail->subject = $GLOBALS['TL_LANG']['MSC']['lockedAccount'][0];
-				$objEmail->text = sprintf($GLOBALS['TL_LANG']['MSC']['lockedAccount'][1], $this->username, ((TL_MODE == 'FE') ? $this->firstname . " " . $this->lastname : $this->name), \Idna::decode(\Environment::get('base')), ceil(\Config::get('lockPeriod') / 60));
-				$objEmail->sendTo(\Config::get('adminEmail'));
-			}
 
-			return false;
-		}
+	/**
+	 * Set all user properties from a database record
+	 */
+	abstract protected function setUserFromDb();
 
-		// Check the account status
-		if ($this->checkAccountStatus() == false)
-		{
-			return false;
-		}
 
-		$blnNeedsRehash = true;
+	/**
+	 * {@inheritdoc}
+	 */
+	abstract public function getRoles();
```

Comment:
This is a BC break. We must not alter the interfaces or abstract classes. We should make the method non-abstract and return a default value (probably an empty array?) instead.

Reply:
🙈  true! Good catch!